src/datastore_mad/remotes/lxd_downloader.sh

```diff
@@ -0,0 +1,161 @@
+#!/bin/bash
+
+get_tag_name () {
+    version=`oned -v | grep "is distributed" | awk '{print $2}'`
+    version_numbers=`echo $version | tr -d .`
+    version_length=`echo $version_numbers | wc -c`
+    if [ $version_length -eq "4" ]; then
+        version_numbers=$(($version_numbers * 10))
+    fi
+        
+    selected_tag="none"
+    for tag in `curl -sSL https://api.github.com/repos/OpenNebula/addon-context-linux/releases | 
+grep "\"tag_name\":" | awk '{print $2}' | cut -d 'v' -f 2 | cut -d '"' -f 1`; do
+        tag_numbers=`echo $tag | tr -d .`
+        tag_length=`echo $tag_numbers | wc -c`
+        if [ $tag_length -eq "4" ]; then
+            tag_numbers=$(($tag_numbers * 10))
+        fi
+        if [ $tag_numbers -le $version_numbers ]; then
+            selected_tag=$tag
+            break
+        fi
+    done
+}
+
+dns_server="8.8.8.8"
+tmp_dir=/var/tmp
+common_dir=/var/tmp/common
+id=`uuidgen`
+url=$1
+url_and_arguments=`echo $url | grep -oP "^"lxd://"\K.*"`
+rootfs_url=`echo $url_and_arguments | cut -d '?' -f 1`
+arguments=`echo $url_and_arguments | cut -d '?' -f 2`
+get_tag_name
+#Create a shell variable for every parameter and# 
+#sets it to the correspondig value              #
+for p in ${arguments//&/ };do 
+    kvp=( ${p/=/ } ); 
+    k=${kvp[0]};v=${kvp[1]};
+    eval $k=$v;
+done
+
+rootfs_url=`echo $url | grep -oP "^"lxd://"\K.*"`
+curl="curl -L"
+if [[ $rootfs_url == *"turnkeylinux.org"* ]]; then
+    untar_options="xvzpf"
+    extension="tar.gz"
+    distro=`echo $rootfs_url | cut -d '-' -f 1 | cut -d '/' -f 7`
+    version_number=`echo $rootfs_url | cut -d '-' -f 2`
+    if [[ $version_number == "9" ]]; then
+        version="Stretch"
+    elif [[ $version_number == "8" ]]; then
+        version="Jessie"
+    else
+        version="undefined"
+    fi
+    release_date=""
+elif [[ $rootfs_url == *"linuxcontainers.org"* ]]; then
+    untar_options="xvJpf"
+    extension="tar.xz"
+    distro=`echo $rootfs_url | cut -d '/' -f 5`
+    version=`echo $rootfs_url | cut -d '/' -f 6`
+    release_date=`echo $rootfs_url | cut -d '/' -f 9`
+fi
+
+output=$tmp_dir/$id.$extension
+
+$curl $rootfs_url --output $output --silent
+qemu-img create -f raw $tmp_dir/$id.raw ${size}M  > /dev/null 2>&1
+
+case $filesystem in
+    "ext4")
+        mkfs.ext4 -F $tmp_dir/$id.raw > /dev/null 2>&1
+        ;;
+    "xfs")
+        mkfs.xfs -f $tmp_dir/$id.raw > /dev/null 2>&1
+        ;;
+    *)
+        mkfs.ext4 -F $tmp_dir/$id.raw > /dev/null 2>&1
+        ;;
+esac
+
+
+mkdir $tmp_dir/$id
+sudo mount $tmp_dir/$id.raw $tmp_dir/$id
```

Comment:
I think there could be a different way if we focus only on ext4. Do all the necessary changes in the chroot and then copy into ext4 raw image directory by directory and file by file via `debugfs` command. Haven't tried it, it's a much harder way, but like it more than mount/sudo chroot... https://www.cs.montana.edu/courses/309/topics/4-disks/debugfs_example.html

Reply:
Haven't tried debugfs before, but, as you said, it only supports ext, and it is a requirement for us to support more filesystems at the moment (like XFS). I tried several FUSE mount tools, having the best results with guestmount but it  was not good: dependencies issues, a daemon (forgot which) must be configured and started, there was even a bug on ubuntu 18.04 that needs to be fixed before using it...so I went back to sudo mount, too many things could go wrong.